File: agent/src/life_update_agent/capture/window_tracker.py
"""Layer 1 - active window/app tracking.

Polls the OS for the active window on an interval and records a row only
when it changes (app or title), so a session spent in one window doesn't
flood the store. Exclude-list is checked before an event is built; whatever
survives is passed through the Layer 2 scanner before being written.
"""

from __future__ import annotations

import logging
import time
from datetime import datetime, timezone

import pywinctl

from life_update_agent import db
from life_update_agent.capture.exclude_list import is_excluded
from life_update_agent.config import ExcludeList
from life_update_agent.redaction.scanner import scan

logger = logging.getLogger(__name__)

POLL_INTERVAL_SECONDS = 2.0


def _read_active_window() -> tuple[str | None, str | None]:
    """Returns (app_name, title), or (None, None) if unavailable (no window
    focused, or the OS denied permission - logged once, never raised)."""
    try:
        window = pywinctl.getActiveWindow()
    except Exception:
        logger.warning("failed to read active window (permissions?)", exc_info=True)
        return None, None

    if window is None:
        return None, None

    try:
        app_name = window.getAppName()
    except Exception:
        app_name = None

    title = getattr(window, "title", None) or None
    return app_name, title


def run(exclude_list: ExcludeList, stop_event) -> None:
    last_seen: tuple[str | None, str | None] | None = None

    while not stop_event.is_set():
        app_name, title = _read_active_window()

        if (app_name, title) != last_seen and (app_name or title):
            last_seen = (app_name, title)

            if not is_excluded(exclude_list, app_name, title):
                with db.get_conn() as conn:
                    conn.execute(
                        "INSERT INTO raw_events (ts, kind, app_name, window_title) VALUES (?, ?, ?, ?)",
                        (
                            datetime.now(timezone.utc).isoformat(),
                            "window",
                            scan(app_name),
                            scan(title),
                        ),
                    )

        stop_event.wait(POLL_INTERVAL_SECONDS)

File: app-shell/src-tauri/src/agent/window_tracker.rs
//! Layer 1 - active window/app tracking. Polls the OS and records a row
//! only when the (app, title) pair changes. Exclude-list is checked before
//! an event is built; whatever survives goes through the Layer 2 scanner.

use super::{db, is_excluded, redaction::scan, AgentConfig};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

const POLL_INTERVAL: Duration = Duration::from_secs(2);

pub fn read_active_window() -> (Option<String>, Option<String>) {
    match active_win_pos_rs::get_active_window() {
        Ok(w) => {
            let app = if w.app_name.is_empty() { None } else { Some(w.app_name) };
            let title = if w.title.is_empty() { None } else { Some(w.title) };
            (app, title)
        }
        Err(_) => (None, None),
    }
}

pub fn run(cfg: Arc<AgentConfig>, stop: Arc<AtomicBool>) {
    let mut last_seen: Option<(Option<String>, Option<String>)> = None;

    while !stop.load(Ordering::Relaxed) {
        let (app_name, title) = read_active_window();

        let current = (app_name.clone(), title.clone());
        if last_seen.as_ref() != Some(&current) && (app_name.is_some() || title.is_some()) {
            last_seen = Some(current);

            if !is_excluded(&cfg, app_name.as_deref(), title.as_deref()) {
                db::insert_raw(
                    &cfg.db_path,
                    &db::now_iso(),
                    "window",
                    scan(app_name.as_deref()).as_deref(),
                    scan(title.as_deref()).as_deref(),
                    None,
                    None,
                );
            }
        }

        std::thread::sleep(POLL_INTERVAL);
    }
}
